corehq/apps/celery/durable.py

```diff
@@ -0,0 +1,83 @@
+import kombu.utils.json as kombu_json
+from celery import Task
+from celery import states as celery_states
+from celery.signals import task_postrun
+from dimagi.utils.logging import notify_exception
+
+
+class UnsupportedSerializationError(Exception):
+    """Raised when attempting to use serialization other than JSON"""
+
+
+class DurableTask(Task):
+    """
+    Only supports tasks configured with json serialization (i.e. no support for pickle)
+    """
+
+    def __init__(self):
+        super().__init__()
+        if getattr(self, 'durable', False):
+            if self.serializer != 'json':
+                raise UnsupportedSerializationError(
+                    f"Cannot create a durable task with {self.serializer} serialization."
+                    "Durable tasks only support JSON serialization."
+                )
+
+    def apply_async(self, args=None, kwargs=None, **opts):
+        from corehq.apps.celery.models import TaskRecord
+
+        if not getattr(self, 'durable', False):
+            return super().apply_async(args=args, kwargs=kwargs, **opts)
+
+        headers = opts.pop('headers', {}) or {}
+        headers['durable'] = True
+
+        record = None
+        task_id = opts.get('task_id', None)
+        if task_id:
+            # if a task calls `self.retry()`, it triggers another apply_async call
+            try:
+                record = TaskRecord.objects.get(task_id=task_id)
+            except TaskRecord.DoesNotExist as e:
+                notify_exception(
+                    None,
+                    f'Unexpectedly could not find TaskRecord object for id {task_id}',
+                    details={'error': str(e)},
+                )
+
+        if not record:
+            record = TaskRecord(
+                name=self.name,
+                args=kombu_json.dumps(args),
+                kwargs=kombu_json.dumps(kwargs),
+                sent=False,
+            )
```

Comment:
Would this be too performance intensive?
```suggestion
            )
        else:
            assert record.name == self.name, (record.name, self.name)
            assert record.args == kombu_json.dumps(args), (record.args, kombu_json.dumps(args))
            assert record.kwargs == kombu_json.dumps(kwargs), (record.kwargs, kombu_json.dumps(kwargs))
```

Reply:
[f822684](https://github.com/dimagi/commcare-hq/pull/37079/commits/f82268417b2031e5019d12338742cdb9761d658d)